[hdmi_cec] new component to communicate with hdmi devices

JosVanEijndhoven opened the pull request:
# What does this implement/fix?

This component allows to connect an esphome device to the 'HDMI CEC' control bus.
On that bus, it can issue and receive commands to/from A/V devices in an HDMI-connected cluster.
This includes commands to power devices on or off, adjust volume, select A/V inputs, etc.

## Types of changes

The 'hdmi_cec' component is new to the esphome repository.
However, earlier versions of this were available (and have been used) from persons own repos:

- https://github.com/johnboiles/esphome-hdmi-cec
- https://github.com/Palakis/esphome-native-hdmi-cec

With respect to the above repos:

- Bugfix: Fix operation on esp32c3
- New feature: Add message decoding into log output for understanding the communication protocol
- New feature: Add optional UART support and buffering of messages, to avoid occasional crashes on all platforms due to excessive operation duration
- Code quality improvements to existing code: Restructuring into more classes to clarify functionality
- Improve inter-thread operation with atomics

Some of those (fix for esp32c3, add message decoding) were also back-ported into Palakis' repo.

For now, the 'message decoder' [is removed](https://github.com/JosVanEijndhoven/esphome/commit/11ded9dfbf74f575f198fbe4ba90d6db7cf42112) to reduce the size of this PR for review. This decoder was optional for use, configured in yaml, and improves the understanding of the bus message logging. The version *with* the decoder can still be obtained from https://github.com/JosVanEijndhoven/esphome/tree/d7a1cd002aa762e12f2af99d866126deee70be1a .

**Related issue or feature (if applicable):**

The component earlier implementation, as available in the developers repos as mentioned above, suffered from long running times per `loop()` invocation: This could range from 100 to 500msec, and lead to device instability and occasional crash&reboot.
This behavior is now improved by buffering and intermediate state for message re-transmissions.

Optionally, this new version can utilize a UART with which the long occupations are fully gone.
This usage is further explained in the document section.

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#6073

## Test Environment

- ESP32s3
- ESP32c3
- ESP32 IDF
- ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example WITH the optional use of a UART:
  uart:
    id: cec_uart
    tx_pin: GPIO7
    baud_rate: 115200

  hdmi_cec:
    pin: GPIO9
    uart_id: cec_uart
    address: 0xE
    physical_address: 0x4000
    on_message:
      - opcode: 0x36 # opcode for "Standby"
        then:
          logger.log: "Got Standby command"
```

## Checklist:
  - [*] The code change is tested and works locally. (tested on esp32s3, esp32c3, esp8266)
  - [*] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [*] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Palakis committed:
initial commit

Palakis committed:
initial commit

Palakis committed:
add pin

Palakis committed:
add pin

Palakis committed:
internal pin

Palakis committed:
internal pin

Palakis committed:
interrupt

Palakis committed:
interrupt

Palakis committed:
interrupts++

Palakis committed:
interrupts++

Palakis committed:
fix

Palakis committed:
fix

Palakis committed:
decoder wip

Palakis committed:
decoder wip

Palakis committed:
fix

Palakis committed:
fix

Palakis committed:
bit shift fix

Palakis committed:
bit shift fix

Palakis committed:
truc

Palakis committed:
truc

Palakis committed:
frame queue

Palakis committed:
frame queue

Palakis committed:
remove logging in interrupts

Palakis committed:
remove logging in interrupts

Palakis committed:
interrupt logging test

Palakis committed:
interrupt logging test

Palakis committed:
fixes

Palakis committed:
fixes

Palakis committed:
interrupt fix

Palakis committed:
interrupt fix

Palakis committed:
better frame logging

Palakis committed:
better frame logging

Palakis committed:
processing loop refactor

Palakis committed:
processing loop refactor

Palakis committed:
decode header

Palakis committed:
decode header

Palakis committed:
address + prom. mode

Palakis committed:
address + prom. mode

Palakis committed:
prom mode in config dump

Palakis committed:
prom mode in config dump

Palakis committed:
nitpick

Palakis committed:
nitpick

Palakis committed:
trigger wip

Palakis committed:
trigger wip

Palakis committed:
process broadcast as well

Palakis committed:
process broadcast as well

Palakis committed:
nitpick check

Palakis committed:
nitpick check

Palakis committed:
logging: improved logging

Palakis committed:
logging: improved logging

Palakis committed:
logging fix

Palakis committed:
logging fix

Palakis committed:
uppercase hex

Palakis committed:
uppercase hex

Palakis committed:
Support for CEC Send + byte ACK (#1)

Palakis committed:
Support for CEC Send + byte ACK (#1)

Palakis committed:
send: wait for "bus free" condition before every attempt

Palakis committed:
send: wait for "bus free" condition before every attempt

Palakis committed:
send: improved logging

Palakis committed:
send: improved logging

Palakis committed:
send: reworked log levels

Palakis committed:
send: reworked log levels

Palakis committed:
quick refactor

Palakis committed:
quick refactor

Palakis committed:
recv+send: fix ack

Palakis committed:
recv+send: fix ack

Palakis committed:
ack: add interrupt lock to preserve timing

Palakis committed:
ack: add interrupt lock to preserve timing

Palakis committed:
automation(send): make destination templatable

Palakis committed:
automation(send): make destination templatable

Palakis committed:
interrupt: don't reset last_falling_edge_us after rising edge

Palakis committed:
interrupt: don't reset last_falling_edge_us after rising edge

Palakis committed:
recv: fix ack only sent for the first byte of every frame

Palakis committed:
recv: fix ack only sent for the first byte of every frame

Palakis committed:
README and LICENSE

Palakis committed:
README and LICENSE

Palakis committed:
fix esp-idf compile error

Palakis committed:
fix esp-idf compile error

Palakis committed:
builtin handlers WIP

Palakis committed:
builtin handlers WIP

Palakis committed:
builtin handler: "Give OSD Name"

Palakis committed:
builtin handler: "Give OSD Name"

Palakis committed:
fix built-in handlers triggering in promiscuous mode

Palakis committed:
fix built-in handlers triggering in promiscuous mode

Palakis committed:
add physical address parameter + "Give Physical Address" handler

Palakis committed:
add physical address parameter + "Give Physical Address" handler

Palakis committed:
fallback handler + ability to override built-in handlers

Palakis committed:
fallback handler + ability to override built-in handlers

Palakis committed:
builtin handler: fix infinite "Feature Abort" loop

Palakis committed:
builtin handler: fix infinite "Feature Abort" loop

Palakis committed:
esp8266 fix

Palakis committed:
esp8266 fix

niko132 committed:
add monitor mode (#4)

niko132 committed:
add monitor mode (#4)

Minims committed:
Broadcast Physical Address & Doc (#5)

* Broadcast Physical Address

Broadcast Physical Address.
My TV does not view the esphome device if it's not broadcast.

* Update README.md to add Services capabilities

Minims committed:
Broadcast Physical Address & Doc (#5)

* Broadcast Physical Address

Broadcast Physical Address.
My TV does not view the esphome device if it's not broadcast.

* Update README.md to add Services capabilities

Palakis committed:
hdmi_cec.send: make "source" option templatable

Fixes #2

Palakis committed:
hdmi_cec.send: make "source" option templatable

Fixes #2

segfly committed:
Added pin setup to fix issues with ESP32-C3 (#8)

Co-authored-by: Nicholas Pace <3300476+segfly@users.noreply.github.com>

segfly committed:
Added pin setup to fix issues with ESP32-C3 (#8)

Co-authored-by: Nicholas Pace <3300476+segfly@users.noreply.github.com>

Palakis committed:
hdmi_cec: expose bytes_to_string

Palakis committed:
hdmi_cec: expose bytes_to_string

Palakis committed:
hdmi_cec: use delay_microseconds_safe everywhere

Palakis committed:
hdmi_cec: use delay_microseconds_safe everywhere

Palakis committed:
hdmi_cec: replace busy-loops with delay_microseconds_safe

Palakis committed:
hdmi_cec: replace busy-loops with delay_microseconds_safe

JosVanEijndhoven committed:
First set of edits in 'hdmi_cec' to create a send queue and prepare for uart operation

JosVanEijndhoven committed:
First set of edits in 'hdmi_cec' to create a send queue and prepare for uart operation

JosVanEijndhoven committed:
First set of edits in 'hdmi_cec' to add UART configuration and use

JosVanEijndhoven committed:
First set of edits in 'hdmi_cec' to add UART configuration and use

JosVanEijndhoven committed:
First successful 'test_cec' both with and without uart

JosVanEijndhoven committed:
First successful 'test_cec' both with and without uart

JosVanEijndhoven committed:
Conditionally include UARTComponent type, as the required include might be absent

JosVanEijndhoven committed:
Conditionally include UARTComponent type, as the required include might be absent

JosVanEijndhoven committed:
Create Message class

JosVanEijndhoven committed:
Create Message class

JosVanEijndhoven committed:
Create Message class

JosVanEijndhoven committed:
Create Message class

JosVanEijndhoven committed:
Introduce CECTransmit class to bundle transmit things

JosVanEijndhoven committed:
Introduce CECTransmit class to bundle transmit things

JosVanEijndhoven committed:
Introduce CECReceive class to bundle receive things

JosVanEijndhoven committed:
Introduce CECReceive class to bundle receive things

JosVanEijndhoven committed:
Move state enums to inside Transmit and Receiver classes, rename consts in capitals

JosVanEijndhoven committed:
Move state enums to inside Transmit and Receiver classes, rename consts in capitals

JosVanEijndhoven committed:
Modify Receiver->Transmitter synchronization to be more robust, avoid deadlock

JosVanEijndhoven committed:
Modify Receiver->Transmitter synchronization to be more robust, avoid deadlock

JosVanEijndhoven committed:
Several small code  enhancements, extra comments, update test

JosVanEijndhoven committed:
Several small code  enhancements, extra comments, update test

JosVanEijndhoven committed:
Receiver sends the 'ack' bit by the uart if available, to reduce isr duration

JosVanEijndhoven committed:
Receiver sends the 'ack' bit by the uart if available, to reduce isr duration

JosVanEijndhoven committed:
rename HAVE_UART #define, protect 'send_ack..' with this define

JosVanEijndhoven committed:
rename HAVE_UART #define, protect 'send_ack..' with this define

JosVanEijndhoven committed:
change some 'const' to 'constexpr', change uart data vector for single write

JosVanEijndhoven committed:
change some 'const' to 'constexpr', change uart data vector for single write

JosVanEijndhoven committed:
small fix

JosVanEijndhoven committed:
small fix

JosVanEijndhoven committed:
move new hdmi-cec component files to proper directory

JosVanEijndhoven committed:
move new hdmi-cec component files to proper directory

JosVanEijndhoven committed:
Add initial 'hdmi_cec' source and test files

JosVanEijndhoven committed:
Add initial 'hdmi_cec' source and test files

JosVanEijndhoven committed:
Modify and extend log messages, fix uart ack bit

JosVanEijndhoven committed:
Modify and extend log messages, fix uart ack bit

JosVanEijndhoven committed:
Add component test files for esp32, esp32-s3, and rp2040

JosVanEijndhoven committed:
Add component test files for esp32, esp32-s3, and rp2040

JosVanEijndhoven committed:
Modify cec pin mode to support 8266, add esp8266 tests

JosVanEijndhoven committed:
Modify cec pin mode to support 8266, add esp8266 tests

JosVanEijndhoven committed:
fix receiving of messages by removing allocation (and std::queue) from isr

JosVanEijndhoven committed:
fix receiving of messages by removing allocation (and std::queue) from isr

JosVanEijndhoven committed:
small update to comment

JosVanEijndhoven committed:
small update to comment

JosVanEijndhoven committed:
Adopt ring buffer for Messages

JosVanEijndhoven committed:
Adopt ring buffer for Messages

JosVanEijndhoven committed:
Add test files for esp32-c3

JosVanEijndhoven committed:
Add test files for esp32-c3

JosVanEijndhoven committed:
Add message decoder in separate include file

JosVanEijndhoven committed:
Add message decoder in separate include file

JosVanEijndhoven committed:
Reimplemented message decoder as class, in files cec_decoder.cpp/.h

JosVanEijndhoven committed:
Reimplemented message decoder as class, in files cec_decoder.cpp/.h

JosVanEijndhoven committed:
Updates to the decoder, rename class Message to Frame

JosVanEijndhoven committed:
Updates to the decoder, rename class Message to Frame

JosVanEijndhoven committed:
Added some code for debugging the status

JosVanEijndhoven committed:
Added some code for debugging the status

JosVanEijndhoven committed:
Add example from esphome-Docs repo for testing

JosVanEijndhoven committed:
Add example from esphome-Docs repo for testing

JosVanEijndhoven committed:
Remove wifi scan info from example log

JosVanEijndhoven committed:
Remove wifi scan info from example log

JosVanEijndhoven committed:
Update to hdmi_cec test of document example

JosVanEijndhoven committed:
Update to hdmi_cec test of document example

JosVanEijndhoven committed:
Update tests for hdmi_cec: add idf

JosVanEijndhoven committed:
Update tests for hdmi_cec: add idf

JosVanEijndhoven committed:
hdmi_cec: small update on log messages

JosVanEijndhoven committed:
hdmi_cec: small update on log messages

JosVanEijndhoven committed:
hdmi_cec: update for code owner, configure use decoder

JosVanEijndhoven committed:
hdmi_cec: update for code owner, configure use decoder

JosVanEijndhoven committed:
hdmi_cec: bug fix cec_decoder after previous edit

JosVanEijndhoven committed:
hdmi_cec: bug fix cec_decoder after previous edit

JosVanEijndhoven committed:
hdmi_cec: edit doc example test

JosVanEijndhoven committed:
hdmi_cec: edit doc example test

JosVanEijndhoven committed:
several tiny fixes to satisfy lint

JosVanEijndhoven committed:
several tiny fixes to satisfy lint

JosVanEijndhoven committed:
small code cleanup in hdmi_cec, extend test

JosVanEijndhoven committed:
small code cleanup in hdmi_cec, extend test

JosVanEijndhoven committed:
Fix clang style complaints

JosVanEijndhoven committed:
Fix clang style complaints

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

JosVanEijndhoven committed:
Remove weird chars from file, adjust formatting

JosVanEijndhoven committed:
Remove weird chars from file, adjust formatting

JosVanEijndhoven committed:
Circumvent atomic increment operations: those fail on esp-c3

JosVanEijndhoven committed:
Circumvent atomic increment operations: those fail on esp-c3

JosVanEijndhoven committed:
Fix complaint on #define

JosVanEijndhoven committed:
Fix complaint on #define

trvrnrth commented on esphome/components/hdmi_cec/cec_decoder.cpp:
```diff
@@ -0,0 +1,462 @@
+}
+
+template<> bool Decoder::do_operand_<Decoder::DISPLAY_CONTROL>() {
+  const static std::array<const char *, 9> NAMES = {"Default Time", "Until cleared", "Clear previous",
```
There are some unicode control characters lurking on here which I imagine was not intentional.

JosVanEijndhoven replied:
Very weird...  I fixed these strange characters.
I also synced and rebased again, which now resolved the earlier issue of recreating CODEOWNERS. So the CI checks pass now :-)

JosVanEijndhoven replied:
Similarly, a new sync-and-rebase fixed an issue with the CI of the document generation. So its preview is now online: https://deploy-preview-5441--esphome.netlify.app/components/hdmi_cec/

JosVanEijndhoven committed:
Fix weird unicode characters

JosVanEijndhoven committed:
Fix weird unicode characters

JosVanEijndhoven committed:
Small update on test of document hdmi_cec example

JosVanEijndhoven committed:
Small update on test of document hdmi_cec example

JosVanEijndhoven committed:
Updated CODEOWNERS for hdmi-cec

JosVanEijndhoven committed:
Updated CODEOWNERS for hdmi-cec

JosVanEijndhoven committed:
Trigger on received message matching on initial part

JosVanEijndhoven committed:
Trigger on received message matching on initial part

JosVanEijndhoven committed:
Update hdmi_cec use of Action::play to new 2025.12 esphome.core version

JosVanEijndhoven committed:
Update hdmi_cec use of Action::play to new 2025.12 esphome.core version

JosVanEijndhoven committed:
hdmi_cec properly uses atomic incr, with fallback for old platform

JosVanEijndhoven committed:
hdmi_cec properly uses atomic incr, with fallback for old platform

JosVanEijndhoven committed:
hdmi_cec fix small formatting issue

JosVanEijndhoven committed:
hdmi_cec fix small formatting issue

trvrnrth commented on esphome/components/hdmi_cec/hdmi_cec.h:
```diff
@@ -0,0 +1,278 @@
+#define ATOMIC_INCR(atom) ((atom)++)
+#else
+// This fallback is probably only for the older esp8266, which does not provide this atomic opration
+#define ATOMIC_INCR(atom) \
```
I have no expertise in the area but can `fetch_add` not be used at all times as seen elsewhere rather than defining this macro?

JosVanEijndhoven replied:
Hi Trevor, thanks for looking into my code!
The atomic ++ operator is translated by gcc into a call to `fetch_add`. However, the older esp8266 library seems to not provide that. Therefor, the use of  `#ifdef` to remove such atomic increment.
If I use either '++' or a direct call to `fetch_add(1u)`, a linker error results only for the esp8266 platform:
```
....xtensa-lx106-elf/bin/ld: hdmi_cec.cpp:(.iram.text.hdmi_cec.cpp.562.34+0x76): undefined reference to `__atomic_fetch_add_1'
```
Because the build infrastructure conditionally provides the now used `ESPHOME_THREAD_MULTI_ATOMICS` define, I suppose that this is expected behavior.

JosVanEijndhoven replied:
I slightly simplified the code for the non-atomic increment, avoiding the unnecessary typecast.

JosVanEijndhoven replied:
Can I ask another question? The 'esphome bot' has one remaining issue, telling this PR is too large. Do I need to do something to bypass that?

trvrnrth replied:
Ah, looking a bit more carefully I see the other uses of `fetch_add` are gated for ESP32 only which I hadn't noticed when commenting.

> Can I ask another question? The 'esphome bot' has one remaining issue, telling this PR is too large. Do I need to do something to bypass that?

I have no authority on the matter but for a new integration the PR size here doesn't seem unreasonable to me.

JosVanEijndhoven committed:
hdmi_cec simplify atomic fallback

JosVanEijndhoven committed:
hdmi_cec simplify atomic fallback

JosVanEijndhoven committed:
hdmi_cec simplify atomic fallback #define

JosVanEijndhoven committed:
hdmi_cec simplify atomic fallback #define

JosVanEijndhoven committed:
Update a few comments, remove last 'TODO's

JosVanEijndhoven committed:
Update a few comments, remove last 'TODO's

swoboda1337 reviewed (changes requested):
Removing the doc_example folder (that I what the docs are for) and all esp32 tests except one would help reduce the PR size. You also should remove the common stuff like wifi and the device name from common_*.yaml (look at other tests).

BTW more tests are not better. We have 100s of components to test, reducing test time is imporant and unless there are variant specific ifdefs only one variant is needed.

JosVanEijndhoven committed:
Remove tests for c3 and 8266 platforms, remove test doc_example

JosVanEijndhoven committed:
Remove tests for c3 and 8266 platforms, remove test doc_example

JosVanEijndhoven committed:
Simplify hdmi_cec tests common*yaml files, run 'no_uart' only on idf

JosVanEijndhoven committed:
Simplify hdmi_cec tests common*yaml files, run 'no_uart' only on idf

JosVanEijndhoven commented:
Hi @swoboda1337 , thanks for your help!
As you requested, I have removed the test files for the 'other' (c3, 8266) platforms, and removed one of the two 'no_uart' tests to further reduce redundancy. The  `doc_example` test dir was also removed. Some initial lines from the `common*yaml` were stripped. However, I left the `wifi:` in one of these files, as I liked to see the `api: services: ` in the test, which seems to require the `wifi:`.  OK?
(These yaml files were somewhat longer as needed for the CI test, to have them actually run on hardware.)

JosVanEijndhoven committed:
Reorganized tests fir 'hdmi-cec', 3 target platforms, no 'wifi:'

JosVanEijndhoven committed:
Reorganized tests fir 'hdmi-cec', 3 target platforms, no 'wifi:'

JosVanEijndhoven committed:
Fix ci error to avoid 'sprintf'

JosVanEijndhoven committed:
Fix ci error to avoid 'sprintf'

JosVanEijndhoven committed:
Fix ci error on 'sprintf', suppress ci on my own 'to_string'

JosVanEijndhoven committed:
Fix ci error on 'sprintf', suppress ci on my own 'to_string'

JosVanEijndhoven committed:
Fix ci error on 'sprintf'

JosVanEijndhoven committed:
Fix ci error on 'sprintf'

bdraco reviewed (commented):
`Frame` inherits from `std::vector<uint8_t>` and gets `push_back()`'d inside `gpio_isr_()` (hdmi_cec.cpp:653). `std::vector::push_back` can trigger heap reallocation inside an interrupt handler which is undefined behavior and a crash waiting to happen.

The `reserve(16)` in the ring buffer constructor makes it unlikely in practice, but nothing enforces that invariant. If a malformed bus signal produces >16 bytes, you get a heap alloc in ISR context.

A fixed-size buffer (`std::array<uint8_t, 16>` + a size field) would eliminate this risk entirely and be more appropriate for ISR use.

bdraco reviewed (commented):
The synchronization model here is mixing up threading and interrupt preemption. The atomics on the ring buffer indices don't help because the actual data being protected, the Frame contents written via `std::vector::push_back` inside the ISR, has no atomic or memory barrier protection at all. On dual core ESP32, the main loop could see an updated ring buffer index but stale Frame data.

The `ATOMIC_INCR` macro with an ESP8266 fallback is a sign the abstraction doesn't fit. For ISR to main loop communication, the standard embedded approach is `volatile` + critical sections (disable/enable interrupts), not `std::atomic`.

bdraco reviewed (commented):
`HAVE_UART` is way too generic for a global define. This could easily collide with other components or system headers. Should be namespaced to the component, something like `USE_HDMI_CEC_UART`.

Also, both `HAVE_UART` and `USE_CEC_DECODER` need to be added to `esphome/core/defines.h` for IDE support and static analysis.

bdraco reviewed (commented):
`std::queue` uses `std::deque` as its default backing container, which allocates in 512 byte blocks. On memory constrained devices this is a lot of RAM for what's essentially a small send buffer.

The component already has `FrameRingBuffer` for the receive side. Could reuse the same pattern here, or use a simple bounded buffer since the send queue shouldn't need to be unbounded anyway.

bdraco reviewed (commented):
`std::map` pulls in the red-black tree template machinery which is pretty heavy on flash. There are three instances here: `CEC_OPCODE_TABLE` (~60 entries), `VENDOR_IDS` (~28 entries), and the implicit one for lookups.

Since these are all const/static lookup tables, a sorted `std::array` with a simple linear scan or binary search would do the same job without the `_Rb_tree` overhead. For the opcode table a switch statement would also work.

bdraco reviewed (commented):
Two small things in `cec_decoder.cpp`: opcode `0x00` ("Feature Abort") appears twice in `CEC_OPCODE_TABLE` (lines 322 and 377), the second one is silently ignored by `std::map`. And in `UI_COMMANDS`, line 67 has `"isplay Information"` instead of `"Display Information"`.

bdraco reviewed (commented):
The `FrameRingBuffer` API is a bit confusing. `push_front()` consumes/pops an element, which is the opposite of what that name usually means. Something like `pop_front()` would be clearer.

Also `back()` calls `clear()` on the returned Frame as a side effect, which is surprising for something that looks like a read accessor. Worth splitting that into an explicit step.

bdraco reviewed (commented):
All the string literals in the decoder tables (`UI_COMMANDS`, `AUDIO_FORMATS`, `AUDIO_SAMPLERATES`, `VENDOR_IDS`, `CEC_OPCODE_TABLE` names, and the various `NAMES` arrays in the operand decoders) will be stored in RAM on ESP8266. That's a lot of strings eating into a very small heap.

The contiguous arrays like `UI_COMMANDS` could use `PROGMEM_STRING_TABLE` from `esphome/core/progmem.h`. The sparse tables would need individual `PROGMEM`/`PSTR()` treatment. Since ESP8266 is a supported target for this component, this should be addressed.

bdraco reviewed (commented):
nit: the `cv.Optional` defaults here use positional arguments instead of the `default=` keyword.
```python
cv.Optional(CONF_PROMISCUOUS_MODE, False)  # should be default=False
cv.Optional(CONF_MONITOR_MODE, False)
cv.Optional(CONF_DECODE_MESSAGES, True)
cv.Optional(CONF_OSD_NAME, "esphome")
```

bdraco reviewed (commented):
The `hdmi_cec.send` action is defined but never exercised in any of the test YAML files. Also there's no test for GPIO only mode on ESP-IDF (only Arduino is covered for the no-uart case).

bdraco reviewed (commented):
nit: a few typos in comments: "Queing" (hdmi_cec.cpp:232) should be "Queuing", "unsignd" (hdmi_cec.h:63), "memmory" (hdmi_cec.h:39), "frane" (hdmi_cec.h:133).

bdraco reviewed (changes requested):
Thanks for putting this together, HDMI CEC is a great addition and it's clear a lot of work went into this. I did a cursory review and left some comments on design issues that should be addressed before a deeper review.

The biggest concerns are around ISR safety (using `std::vector` inside an interrupt handler, the atomics model), memory overhead (`std::map`, `std::deque`, all the decoder strings landing in RAM on ESP8266), and the `HAVE_UART` define needing to be namespaced.

I'd also suggest splitting this up. The decoder is a self contained piece that could land separately, and the core CEC send/receive logic could be a second PR. At this size it's hard to give a thorough review without a huge time commitment, and smaller PRs tend to move much faster.

bdraco reviewed (commented):
Several functions take `const std::vector<uint8_t> &` where `const uint8_t *data, size_t len` would avoid forcing heap allocations on every call. For example `send()`, `try_builtin_handler_()`, and the `MessageTrigger::set_data()` / trigger signature all pass vectors around. The builtin handler responses (lines 176-216) each construct a temporary `std::vector` just to pass a few bytes.

JosVanEijndhoven committed:
Avoid use of vector push_back in isr

JosVanEijndhoven committed:
Avoid use of vector push_back in isr

JosVanEijndhoven committed:
Frame does NOT inherit from std::array, avoiding bad use of iterators

JosVanEijndhoven committed:
Frame does NOT inherit from std::array, avoiding bad use of iterators

JosVanEijndhoven committed:
Remove unneeded std::move

JosVanEijndhoven committed:
Remove unneeded std::move

JosVanEijndhoven committed:
Fix unsigned vs int comparison

JosVanEijndhoven committed:
Fix unsigned vs int comparison

JosVanEijndhoven committed:
Rename global #defines with HDMI_CEC scope

JosVanEijndhoven committed:
Rename global #defines with HDMI_CEC scope

JosVanEijndhoven commented:
> `HAVE_UART` is way too generic for a global define. This could easily collide with other components or system headers. Should be namespaced to the component, something like `USE_HDMI_CEC_UART`.
> 
> Also, both `HAVE_UART` and `USE_CEC_DECODER` need to be added to `esphome/core/defines.h` for IDE support and static analysis.

Good, the names of these two defines were fixed with an 'HDMI_CEC' prefix as of https://github.com/esphome/esphome/pull/9189/changes/e4797bd8c8339f8672d0409d6bacc113141ae92b

I understand that these become conditionally and automatically added into a project local `esphome/core/defines.h` by the calls to `cg.add_define(` in my component `__init__.py`.

JosVanEijndhoven committed:
Remove std::queue for send. Instead, re-use FrameRingBuffer

JosVanEijndhoven committed:
Remove std::queue for send. Instead, re-use FrameRingBuffer

JosVanEijndhoven commented:
> `std::queue` uses `std::deque` as its default backing container, which allocates in 512 byte blocks. On memory constrained devices this is a lot of RAM for what's essentially a small send buffer.
> 
> The component already has `FrameRingBuffer` for the receive side. Could reuse the same pattern here, or use a simple bounded buffer since the send queue shouldn't need to be unbounded anyway.

I am not very worried about 512 bytes. But indeed, the 'FrameRingBuffer' can very well take (also) this  queue: in practice the number of queued 'sends' would remain very small. Also the 'FrameRingBuffer' has a lock-free design, relying on std::atomic, and hence the std::mutex and LockGuard can also be dropped.
Good suggestion, update was pushed with:
https://github.com/esphome/esphome/pull/9189/changes/608b9bd8385b559229fb24af69c5aa59e544eacc

JosVanEijndhoven committed:
Remove std::map for a local/simple implementation for compactness

JosVanEijndhoven committed:
Remove std::map for a local/simple implementation for compactness

JosVanEijndhoven committed:
Fix typo in UI command strings

JosVanEijndhoven committed:
Fix typo in UI command strings

JosVanEijndhoven commented:
> `std::map` pulls in the red-black tree template machinery which is pretty heavy on flash. There are three instances here: `CEC_OPCODE_TABLE` (~60 entries), `VENDOR_IDS` (~28 entries), and the implicit one for lookups.
> 
> Since these are all const/static lookup tables, a sorted `std::array` with a simple linear scan or binary search would do the same job without the `_Rb_tree` overhead. For the opcode table a switch statement would also work.

OK, I replaced both the `CEC_OPCODE_TABLE` and  `VENDOR_IDS` with a local/simple Lookuptable:
https://github.com/esphome/esphome/pull/9189/changes/373b45059a686bfcbc17d7a8e4862870a8ca7b67

Removing the use of std::map and (from the earlier feedback) std::queue made my test-application indeed _slightly_ smaller.  (a test-app with wifi and other things...) The esphome linker says:
```
             Before     Now        Difference
Flash Code   655104     653224     -1880
Flash Data   149924     149712     -212
RAM:          35904     35856      -48
Flash:       895451     893315     -2136

JosVanEijndhoven commented:
> The synchronization model here is mixing up threading and interrupt preemption. The atomics on the ring buffer indices don't help because the actual data being protected, the Frame contents written via `std::vector::push_back` inside the ISR, has no atomic or memory barrier protection at all. On dual core ESP32, the main loop could see an updated ring buffer index but stale Frame data.
> 
> The `ATOMIC_INCR` macro with an ESP8266 fallback is a sign the abstraction doesn't fit. For ISR to main loop communication, the standard embedded approach is `volatile` + critical sections (disable/enable interrupts), not `std::atomic`.

Here, I do not agree with you, or we somehow misunderstand each other...
The 'FrameRingBuffer' was designed as a 'lock-free' datastructure, where the use of 'std::atomic' enforces memory ordering:

1. In `CECReceive::gpio_isr_()` a gathered byte is written in a pre-allocated `Frame` buffer
2. On an `EOM_ACK`, that Frame buffer is committed by `frames_queue_.push_back()`, which is a write into the atomic `back_inx_`. This establishes a memory-order restriction on the writes into the Frame data.
3. In `HDMICEC::loop()` we get the `recv_.frames_queue_.front()`, which checks `is_empty()`, which reads the atomic value of `back_inx_`.
4. Only after that, we read the frame content in `handle_received_message_()`: so a memory order is enforced as that occurs after reading the atomic.

If I didn't make a mistake here, the memory order is guaranteed by the C++11 semantics, which will work independently for 'real' threads and/or ISR functions (and does not rely on mutexes or critical sections).

It is somewhat shameful that the ESP8266 compiler does not provide the `std::atomic` library. Presumably that is not needed for this single-core processor with a simple instruction pipeline to achieve a decent memory-order behavior.

Jos

JosVanEijndhoven committed:
clarify framebuffer with 'pop_front' and remove 'clear' side effect

JosVanEijndhoven committed:
clarify framebuffer with 'pop_front' and remove 'clear' side effect

JosVanEijndhoven commented:
> The `FrameRingBuffer` API is a bit confusing. `push_front()` consumes/pops an element, which is the opposite of what that name usually means. Something like `pop_front()` would be clearer.
> 
> Also `back()` calls `clear()` on the returned Frame as a side effect, which is surprising for something that looks like a read accessor. Worth splitting that into an explicit step.

OK, fixed with https://github.com/esphome/esphome/pull/9189/changes/16b986f8c528cddd414a2a390c1c0170525e5564

JosVanEijndhoven committed:
Cleanup in cec_decoder

JosVanEijndhoven committed:
Cleanup in cec_decoder

JosVanEijndhoven committed:
fix static constexpr uppercase

JosVanEijndhoven committed:
fix static constexpr uppercase

widget commented:
This is running for me on a SLWF-08 (so no UART) if it helps. Only got some toy examples in at the moment, but I can see HDMI traffic.

I can suggest some slight changes to the message dump or documentation, as it took me a while to understand the first octet is a source nibble **and** destination nibble.

iay commented:
@widget

> This is running for me on a SLWF-08 (so no UART) if it helps.

I happen to have an SLWF-08 recently arrived. Haven't had time to plug it in yet or review the example configuration; hoping I will be able to use this integration when it's done instead.

Can I ask why you're saying that SLWF-08 implies no UART? Is this an ESP8266 vs. ESP32 thing, or an Arduino vs. ESP-IDF) thing, or something else?  Mainly curious.

JosVanEijndhoven committed:
Adjust use of atomics to compile equally on target platforms, adjust atomic usage for recv->xmit signalling

JosVanEijndhoven committed:
Adjust use of atomics to compile equally on target platforms, adjust atomic usage for recv->xmit signalling

JosVanEijndhoven committed:
Remove 'get_state()' debugging'

JosVanEijndhoven committed:
Remove 'get_state()' debugging'

JosVanEijndhoven committed:
Reduce use of std::vector inside hdmi-cec component, in favor of std::initializer_list

JosVanEijndhoven committed:
Reduce use of std::vector inside hdmi-cec component, in favor of std::initializer_list

JosVanEijndhoven committed:
fix for clang CI check

JosVanEijndhoven committed:
fix for clang CI check

JosVanEijndhoven committed:
fix for clang CI check

JosVanEijndhoven committed:
fix for clang CI check

JosVanEijndhoven committed:
Move strings into Flash for 8266 target to save RAM

JosVanEijndhoven committed:
Move strings into Flash for 8266 target to save RAM

JosVanEijndhoven committed:
Revert "Move strings into Flash for 8266 target to save RAM"

This reverts commit 8aa1ef5d0b49ea64a87535a7188d53cdcdd1c284.

JosVanEijndhoven committed:
Revert "Move strings into Flash for 8266 target to save RAM"

This reverts commit 8aa1ef5d0b49ea64a87535a7188d53cdcdd1c284.

JosVanEijndhoven committed:
Add #ifdef for conditional content of 'cec_decoder'

JosVanEijndhoven committed:
Add #ifdef for conditional content of 'cec_decoder'

JosVanEijndhoven committed:
Add 'default=' in 'cv.Optional()'

JosVanEijndhoven committed:
Add 'default=' in 'cv.Optional()'

JosVanEijndhoven committed:
Fix function inlining in IRAM_ATTR methods

JosVanEijndhoven committed:
Fix function inlining in IRAM_ATTR methods

JosVanEijndhoven committed:
Extend tests with 'esp32-s3-idf' and extra 'send'

JosVanEijndhoven committed:
Extend tests with 'esp32-s3-idf' and extra 'send'

JosVanEijndhoven committed:
Fix optional 'source' field in python action

JosVanEijndhoven committed:
Fix optional 'source' field in python action

JosVanEijndhoven committed:
Fix uint32 wrap-around bug on use of 'micros()'

JosVanEijndhoven committed:
Fix uint32 wrap-around bug on use of 'micros()'

JosVanEijndhoven committed:
Lower component setup priority

JosVanEijndhoven committed:
Lower component setup priority

JosVanEijndhoven committed:
New fix to avoid uint32 wraparound issues with 'micros()'

JosVanEijndhoven committed:
New fix to avoid uint32 wraparound issues with 'micros()'

JosVanEijndhoven committed:
Fix order in cec_opcode_table

JosVanEijndhoven committed:
Fix order in cec_opcode_table

JosVanEijndhoven committed:
Merge branch 'dev' into hdmi-cec

JosVanEijndhoven committed:
Small comment update

JosVanEijndhoven committed:
Small code updates from review around python-to-c interface

JosVanEijndhoven committed:
fix typo

JosVanEijndhoven committed:
fix new nested namespace syntax

JosVanEijndhoven committed:
Improve bus busy behavior, fix its timing

JosVanEijndhoven committed:
Modify the message decoder to avoid std::string

JosVanEijndhoven committed:
Merge branch 'dev' into hdmi-cec

JosVanEijndhoven committed:
Merge branch 'dev' into hdmi-cec

JosVanEijndhoven committed:
Remove the decoder sources to reduce the PR review size
